File: gui/linkers/modules/utils/htmlTreeStructure.js
import { ArticleTagCounter } from "./dataTypes.js"
import { correspondingTags } from "./newsUtils.js"

class Node{
    constructor(tag, tagClass=null){
        this.tag = tag
        this.tagClass = tagClass? tagClass: null
    }

    evaluate = (interpreterEnvironment, articleTagCounter) => {}

    getTagClass = () => this.tagClass
    
    countContentTags = () => {}
    
    compare = (node) => {
        if(this.getTagClass === node.getTagClass && this.countContentTags === node.countContentTags) return true
        else return false
    }

    to_s = (start=0) => `${start*" "}${this.tag} (${this.tagClass? this.tagClass: ""})\n`
}

class NonUnaryTagNode extends Node{
    constructor(tag, tagClass=null){
        super(tag, tagClass)
        this.children = []
        this._containsText = null
    }

    addChild = (child) => {
        this.children.push(child)
        this._containsText = null
    }
    
    numChildren = () => this.children.length
    
    popChildren = () => {
        this.children.pop()
        this._containsText = null   
    }

    countContentTags = () => {
        if(!this._containsText){
            if(this.numChildren() === 0) return 1
            else{
                let count = 1
                for(let i=this.children.length-1; i>=0; i--){
                    count += this.children[i].countContentTags()
                }
                this._containsText = count
            }
        }
        return this._containsText
    }
    
    to_s = (start=0, typeTag=null) => {
        let tabs = start * " "
        let tagClass = this.getTagClass()? `(${this.getTagClass()})`: null
        let string = `${tabs}${typeTag? typeTag: ""}${this.tag} ${tagClass}\n`

        if(this.numChildren()>0) children = this.children.map(child => child.to_s(start+1)).join("")
        else children = ""

        return `${string}${children}`
    }
}

class DivNode extends NonUnaryTagNode{
    constructor(tagClass=null){
        super("div", tagClass)
    }
    
    to_s = (start=0) => super.to_s(start, "div")
    
    simple_evaluation = (interpreterEnvironment) => {
        let length = this.children.length
        for(let i=0; i<length; i++){
            this.children[i].simple_evaluation(interpreterEnvironment)
        }
    }

    evaluate = (interpreterEnvironment, articleTagCounter) => {
        if(interpreterEnvironment.articleClasses.has(this.tagClass)){
            interpreterEnvironment.articleClasses.add(this.tagClass)
            articleTagCounter.clear()
        }else{
            let childrenLength = this.children.length
            articleTagCounter.add(2)

            if(childrenLength === 1){
                this.children[0].evaluate(interpreterEnvironment, articleTagCounter)
            }else{
                let newArticleCounter = new ArticleTagCounter()

                for(let i=0; i<childrenLength; i++){
                    let childArticleCounter = new ArticleTagCounter()
                    this.children[i].evaluate(interpreterEnvironment, childArticleCounter)
                    newArticleCounter.merge(childArticleCounter)
                }

                // if(this.tagClass === "l-col__sidebar"){
                //     console.log(this)
                //     console.log(newArticleCounter, newArticleCounter.verify(interpreterEnvironment.decisionTree, interpreterEnvironment.acceptancePercentage),
                //     newArticleCounter.lazyCount(interpreterEnvironment.maxTagsForAcceptance, interpreterEnvironment.minTagsForAcceptance), newArticleCounter.values[3]>0)
                // }

                if(this.tagClass && ((newArticleCounter.values[2] || 0)<10) && newArticleCounter.verify(interpreterEnvironment.decisionTree, interpreterEnvironment.acceptancePercentage)
                    && newArticleCounter.lazyCount(interpreterEnvironment.maxTagsForAcceptance, interpreterEnvironment.minTagsForAcceptance)){
                    // console.log(newArticleCounter.values, newArticleCounter.size)
                    interpreterEnvironment.articleClasses.add(this.tagClass)
                    articleTagCounter.clear()
                }else{
                    articleTagCounter.merge(newArticleCounter)
                }
            }
        }
    }
}
        
class TagNode extends NonUnaryTagNode{
    constructor(tag){
        super(tag)
    }
    
    getTagClass = () => null
    
    evaluate = (interpreterEnvironment, articleTagCounter) => {
        let childrenLength =  this.numChildren()
        if (childrenLength === 0) articleTagCounter.add(correspondingTags[this.tag])
        else{
            for(let i=0; i<childrenLength; i++){
                this.children[i].evaluate(interpreterEnvironment, articleTagCounter)
            }
            
            articleTagCounter.add(correspondingTags[this.tag])
        }
    }
}

class UnaryTagNode extends Node{
    constructor(tag){
        super(tag)
    }
    
    evaluate = (interpreterEnvironment, articleTagCounter) => {articleTagCounter.add(correspondingTags[this.tag])}

    to_s = (start=0) => `${start*" "}${this.tag}\n`

    countContentTags = () => 1

    numChildren = () => 0
    
    addChild = (child) => {}

    popChildren = () => {}
    
    getTagClass = () => null
}

class SimpleNode{
    constructor(tag, tagClass=null, content=null){
        this.tag = tag
        this.tagClass = tagClass
        this.content = content
        this.link = null
    }
    
    count_nodes = () => 1
    
    to_s = (start=0) => `${start*" "}${this.tag}${this.tagClass?this.tagClass:""} -> ${this.content}\n`
    
    // __repr__():
    //     return "%s (%s)" % (.tag, this.tagClass)
    
    getChildren = () => []
    
    isLeaf = () => true
    
    compact = () => {}
}

class SimpleNonUnaryNode extends SimpleNode{
    constructor(tag, tagClass=null, content=null){
        super(tag, tagClass, content)
        this.children = []
    }
    
    addChild = (node) => {this.children.push(node)}
    
    getChildren = () => this.children
    
    isLeaf = () => this.children.length === 0

    count_nodes = () => {
        let sum = 1 

        for(let i=this.children.length-1; i>=0; i--) sum += this.children[i].count_nodes()
        
        return sum
    }
    
    compact = () => {
        let length = this.children.length
        for(let i=0; i<length; i++){
            let child = this.children[i]
            child.compact()
            this.content = child.content
            this.link = child.link
        }
        
        this.children = []
    }
    
    to_s = (start=0) => {
        let text = super.to_s(start)
        let children = (this.count_nodes()-1)>0?this.children.map(child => child.to_s(start+1)).join(""): ""

        return `${text}${children}`
    }
}

class SimpleDivNode extends SimpleNonUnaryNode{
    constructor(tagClass=null, content=null){
        super("div", tagClass, content)
    }
}

class SimpleUnaryTagNode extends SimpleNode{}

class SimpleTagNode extends SimpleNonUnaryNode{}

export { Node, NonUnaryTagNode, DivNode, TagNode, UnaryTagNode, SimpleNode, 
    SimpleNonUnaryNode, SimpleDivNode, SimpleUnaryTagNode, SimpleTagNode }

File: engine/News/htmlTreeStructure.py
from functools import reduce
from Utils.dataTypes import ArticleTagCounter, Counter, InterpreterEnvironment

class Node:
    def __init__(self, tag, tagClass=None):
        self.tag = tag
        self.tagClass = None if tagClass == "" else tagClass

    def evaluate(self, interpreterEnvironment, articleTagCounter):
        pass

    def getTagClass(self):
        return self.tagClass
    
    def countContentTags(self):
        pass
    
    def compare(self, node):
        if(self.getTagClass == node.getTagClass and self.countContentTags == node.countContentTags):
            return True
        else: return False

    def to_s(self, start=0):
        return "%s%s%s\n" % (start*" ", self.tag, f" ({self.tagClass})" if self.tagClass is not None else "")

class NonUnaryTagNode(Node):
    def __init__(self, tag, tagClass=None):
        super().__init__(tag, tagClass)
        self.children = []
        self._containsText = None

    def addChild(self, child):
        self.children.append(child)
        self._containsText = None
    
    def numChildren(self):
        return len(self.children)
    
    def popChildren(self):
        self.children.pop()
        self._containsText = None

    def countContentTags(self):
        if(self._containsText is None):
            if(self.numChildren() == 0): return 1
            else: self._containsText = reduce((lambda x, y: x+y), [child.countContentTags() for child in self.children])

        return self._containsText
    
    def to_s(self, start=0):
        tabs = start * " "
        tagClass = f"({self.getTagClass()})" if self.getTagClass() is not None else ""
        string = f"{tabs}{self.tag} {tagClass}\n"
        if(self.numChildren()>0):
            children = "".join([child.to_s(start+1) for child in self.children])
        else:
            children = ""

        return f"{string}{children}"

class DivNode(NonUnaryTagNode):
    def __init__(self, tagClass=None):
        super().__init__("div", tagClass)
        self.children = []
        self._containsText = None
    
    def to_s(self, start=0):
        tabs = start * " "
        tagClass = f"({self.getTagClass()})" if self.getTagClass() is not None else ""
        string = f"{tabs}div {tagClass}\n"
        if(self.numChildren()>0):
            children = "".join([child.to_s(start+1) for child in self.children])
        else:
            children = ""

        return f"{string}{children}"
    
    def simple_evaluation(self, interpreterEnvironment):
        for child in self.children:
            child.simple_evaluation(interpreterEnvironment)

    def evaluate(self, interpreterEnvironment, articleTagCounter):
        if(self.tagClass in interpreterEnvironment.articleClasses):
            interpreterEnvironment.articleClasses.add(self.tagClass)
            articleTagCounter.clear()
        else:
            if(len(self.children) == 1):
                self.children[0].evaluate(interpreterEnvironment, articleTagCounter)
            else:
                newArticleCounter = ArticleTagCounter()

                for child in self.children:
                    childArticleCounter = ArticleTagCounter()
                    child.evaluate(interpreterEnvironment, childArticleCounter)
                    newArticleCounter.merge(childArticleCounter)

                if(len(newArticleCounter.keys()) == 0): newArticleCounter.add("div")

                if(newArticleCounter.verify(interpreterEnvironment.minTagsForAcceptance, interpreterEnvironment.maxTagsForAcceptance)):
                    #print(self.tagClass, newArticleCounter)
                    interpreterEnvironment.articleClasses.add(self.tagClass)
                    articleTagCounter.clear()
                else: articleTagCounter.merge(newArticleCounter)
        
class TagNode(NonUnaryTagNode):
    def __init__(self, tag):
        self.tag = tag
        self.children = []
        self._containsText = None
    
    def getTagClass(self):
        return None
    
    def evaluate(self, interpreterEnvironment, articleTagCounter):
        if (self.numChildren() == 0):
            articleTagCounter.add(self.tag)
        else:
            for child in self.children:
                child.evaluate(interpreterEnvironment, articleTagCounter)
            
            if(self.tag[0] == "h"): articleTagCounter.add(self.tag)

class UnaryTagNode(Node):
    def __init__(self, tag):
        self.tag = tag
    
    def evaluate(self, interpreterEnvironment, articleTagCounter):
        articleTagCounter.add(self.tag)

    def to_s(self, start=0):
        return "%s%s\n" % (start*" ", self.tag)

    def countContentTags(self):
        return 1

    def numChildren(self):
        return 0
    
    def addChild(self, child):
        return
    
    def popChildren(self):
        return
    
    def getTagClass(self):
        return None

class SimpleNode:
    def __init__(self, tag, tagClass=None, content=None):
        self.tag = tag
        self.tagClass = tagClass
        self.content = content
        self.link = None
    
    def count_nodes(self):
        return 1
    
    def to_s(self, start=0):
        return "%s%s%s -> %s\n" % (start*" ", self.tag, f" ({self.tagClass})" if self.tagClass is not None else "", self.content)
    
    def __repr__(self):
        return "%s (%s)" % (self.tag, self.tagClass)
    
    def getChildren(self):
        return []
    
    def isLeaf(self):
        return True
    
    def compact(self):
        return

class SimpleNonUnaryNode(SimpleNode):
    def __init__(self, tag, tagClass=None, content=None):
        super().__init__(tag, tagClass, content)
        self.children = []
    
    def addChild(self, node):
        self.children.append(node)
    
    def getChildren(self):
        return self.children
    
    def isLeaf(self):
        return (len(self.children) == 0)

    def count_nodes(self):
        sum = 1 

        for child in self.children:
            sum += child.count_nodes()
        
        return sum
    
    def compact(self):
        for child in self.children:
            child.compact()
            self.content = child.content
            self.link = child.link
        
        self.children = []
    
    def to_s(self, start=0):
        tabs = start * " "
        tagClass = f"({self.tagClass})" if self.tagClass is not None else ""
        string = f"{tabs}{self.tag} {tagClass} -> {self.content}\n"
        if(self.count_nodes()-1>0):
            children = "".join([child.to_s(start+1) for child in self.children])
        else:
            children = ""

        return f"{string}{children}"

class SimpleDivNode(SimpleNonUnaryNode):
    def __init__(self, tagClass=None, content=None):
        super().__init__("div", tagClass, content)
    
    def to_s(self, start=0):
        tabs = start * " "
        tagClass = f"({self.tagClass})" if self.tagClass is not None else ""
        string = f"{tabs}div {tagClass} -> {self.content}\n"
        if(self.count_nodes()-1>0):
            children = "".join([child.to_s(start+1) for child in self.children])
        else:
            children = ""

        return f"{string}{children}"
    

class SimpleUnaryTagNode(SimpleNode):
    pass

class SimpleTagNode(SimpleNonUnaryNode):
    pass